### PR TITLE
feat: open source the event generator gradle plugin

### PR DESCRIPTION
**Pull request by jplaisted**
This plugin exists internally and I am now open sourcing it.

I've added two commits to this PR to make reviewing easier. The second is new documentation that should be thoroughly reviewed; while the code changes could perhaps be more skimmed over since they already exist internally.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: gradle-plugins/metadata-annotations-lib/src/main/java/com/linkedin/metadata/annotations/DefaultGmaEntitiesAnnotationAllowList.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import javax.annotation.Nonnull;
+
+/**
+ * Simple allow list for the {@code @gma.aspect.entities} annotation.
+ *
+ * <p>This annotation is only used for helping us migrate old "common" aspects to v5 and should not be used for anything
+ * new in v5.
+ */
+final class DefaultGmaEntitiesAnnotationAllowList implements GmaEntitiesAnnotationAllowList {
+  static final GmaEntitiesAnnotationAllowList DEFAULT =
```
Comment (RealChrisL) [inline gradle-plugins/metadata-annotations-lib/src/main/java/com/linkedin/metadata/annotations/DefaultGmaEntitiesAnnotationAllowList.java:17]: Do we need to expose all the entities in the allowlist, some of them are not open-sourced.

File: docs/what/mxev5.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+**In development.** Check the [roadmap](../roadmap.md) for more details.
+
+## Background
+
+Please read about the current ("v4") events [here](./mxe.md) to get a general understanding of the types of events and
+what they do.
+
+## What are "v5" Metadata Events?
+
+In brief, we've split the monolithic MCE/MAE/FMCE events into smaller events. One event per aspect-entity combination.
```
Comment (RealChrisL) [inline docs/what/mxev5.md:14]: use the order of entity-aspect?

File: docs/what/mxev5.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+This means that MCE transitively references all possible metadata snapshots / aspects in the ecosystem (and similarly
+for MAE/FMCE). We found this to have the follow problems:
+
+- The MXEs referencing all possible models means all models must be in the same project, which:
+  - Made it hard to determine what information you needed.
+  - Makes distrubted ownership of models is not possible.
+    - Internally, this means our team must review all model changes, as only we have review permissions.
+    - Externally (open source), this means that people need to fork the repo to add their own models and need to rebuild
+      the entire project. This can lead to merge conflicts.
+  - We've found internally that this also means that "we're all in this together"; if any team makes a mistake with
```
Comment (RealChrisL) [inline docs/what/mxev5.md:43]: Nice quote!

File: gradle-plugins/metadata-annotations-lib/src/main/java/com/linkedin/metadata/annotations/DefaultGmaEntitiesAnnotationAllowList.java (hunk 1 above)
Comment (jplaisted) [reply gradle-plugins/metadata-annotations-lib/src/main/java/com/linkedin/metadata/annotations/DefaultGmaEntitiesAnnotationAllowList.java:17]: That's a good point. We will end up using this same code, so I'll need to think this through.

File: docs/what/mxev5.md (hunk 3 above)
Comment (jplaisted) [reply docs/what/mxev5.md:43]: I for sure heard it from someone else lol

File: docs/what/mxev5.md (hunk 2 above)
Comment (jplaisted) [reply docs/what/mxev5.md:14]: Any particular reason why?

File: gradle-plugins/metadata-annotations-lib/src/main/java/com/linkedin/metadata/annotations/DefaultGmaEntitiesAnnotationAllowList.java (hunk 1 above)
Comment (jplaisted) [reply gradle-plugins/metadata-annotations-lib/src/main/java/com/linkedin/metadata/annotations/DefaultGmaEntitiesAnnotationAllowList.java:17]: Ok lets do this I guess:

- Make this the list of open source entities to allow.
- Internally, have a separate plugin that applies the open source plugin, but replaces the allow list with the internal allow list. Then internally we apply that.

I'll make another commit to make sure that is extensible.

File: docs/what/mxev5.md (hunk 2 above)
Comment (RealChrisL) [reply docs/what/mxev5.md:14]: because we break down the event in the order of form? `namespace com.linkedin.mxe.corpUser.corpUserInfo`

Comment (jplaisted) [reply docs/what/mxev5.md:14]: The text here isn't meant to indicate order, just the pairing. If you feel it is important to be consistent we can change it.

**Review by RealChrisL — APPROVED** [review]
Thanks for the work!

**Review by woodyzhou — APPROVED** [review]
LGTM

**Pull request merged**